File: go/main/src/worker.go
package main

import (
	"context"
	"log"
	"sync"
	"time"
)

// Worker processes tasks from the tasks channel and writes results to the results channel
func Worker(ctx context.Context, tasks <-chan Task, results chan<- Result, wg *sync.WaitGroup) {
	defer wg.Done() // Mark this worker as done when it finishes

	for {
		select {
		case <-ctx.Done(): // Handle cancellation signal
			log.Println("Worker shutting down due to context cancellation")
			return
		case task, ok := <-tasks: // Retrieve a task from the tasks channel
			if !ok { // Exit if the channel is closed and no more tasks are available
				return
			}

			log.Printf("Worker processing %s\n", task.Name)
			result := processTask(task)
			results <- result // Send the result to the results channel
			log.Printf("Worker completed %s\n", task.Name)
		}
	}
}

// Simulate task processing with a delay and return a result
func processTask(task Task) Result {
	time.Sleep(1 * time.Second) // Simulate processing delay
	return Result{TaskName: task.Name, Status: "Success"}
}

File: java/main/src/Worker.java
import java.util.concurrent.*;
import java.util.logging.*;

public class Worker implements Runnable {

    private final BlockingQueue<Task> taskQueue;
    private final ConcurrentLinkedQueue<Result> results;
    private static final Logger logger = Logger.getLogger(Worker.class.getName());

    public Worker(BlockingQueue<Task> taskQueue, ConcurrentLinkedQueue<Result> results) {
        this.taskQueue = taskQueue;
        this.results = results;
    }

    @Override
    public void run() {
        while (!Thread.currentThread().isInterrupted()) {
            try {
                // Retrieve and process a task from the queue
                Task task = taskQueue.poll(2, TimeUnit.SECONDS);
                if (task == null) {
                    break; // Exit if no more tasks are available
                }

                logger.info(Thread.currentThread().getName() + " processing " + task.getName());
                processTask(task);

                // Add the result to the shared resource
                results.add(new Result(task.getName(), "Success"));
                logger.info(Thread.currentThread().getName() + " completed " + task.getName());

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                logger.severe(Thread.currentThread().getName() + " interrupted: " + e.getMessage());
            } catch (Exception e) {
                logger.severe(Thread.currentThread().getName() + " encountered an error: " + e.getMessage());
            }
        }
    }

    // Simulate task processing with a delay
    private void processTask(Task task) throws InterruptedException {
        Thread.sleep(1000); // Simulate processing delay
    }
}
